SwiftCompilerSources/Sources/Optimizer/Utilities/AccessUtils.swift

```diff
@@ -0,0 +1,421 @@
+         baseAddress.type.isAddress {
+        baseKind = .yield
+      } else {
+        return nil
+      }
+    }
+
+    self.baseAddress = baseAddress
+  }
+
+  init(baseAddress: Value, baseKind: AccessBaseKind) {
+    self.baseAddress = baseAddress
+    self.baseKind = baseKind
+  }
+
+  var description: String {
+    "\(baseKind) - \(baseAddress)"
+  }
+
+  var isObjectAccess: Bool {
+    return baseKind == .class || baseKind == .tail
+  }
+
+  // If this `AccessBase` is obtained froma a reference type,
+  // results in the reference `ref` and a `component` describing
+  // the projection type (tail or class field) and the corresponding field index
+  var referenceWithPathComponent: (ref: Value, component: (SmallProjectionPath.FieldKind, Int))? {
+    switch baseAddress {
+    case let rea as RefElementAddrInst:
+      return (rea.operand, (.classField, rea.fieldIndex))
+    case let pb as ProjectBoxInst:
+      return (pb.operand, (.classField, pb.fieldIndex))
+    case let rta as RefTailAddrInst:
+      return (rta.operand,(.tailElements, 0))
+    default:
+      return nil
+    }
+  }
+
+  var reference: Value? { referenceWithPathComponent?.ref }
+
+  var isLet: Bool {
+    switch baseAddress {
+    case let rea as RefElementAddrInst:
+      return rea.fieldIsLet
+    case let ga as GlobalAddrInst:
+      return ga.global.isLet
+    default:
+      return false
+    }
+  }
+
+  /// Returns `true` if the address is immediately produced by a stack or box allocation
+  var isLocal: Bool {
+    switch baseKind {
+    case .box:
+      // The operand of the projection can be an argument, in which
+      // case it wouldn't be local
+      return (baseAddress as! ProjectBoxInst).operand is AllocBoxInst
+    case .stack:
+      return true
+    default:
+      return false
+    }
+  }
+
+  /// Returns `true` if we can reliably compare for equality this `AccessBase`
+  /// with another `AccessBase`.
+  /// When comparing two uniquely identified access bases, it can be followed that if they are not equal,
+  /// also the accessed memory addresses do not alias.
+  /// This is e.g. not the case for class references: two different references may still point to the same object.
+  var isUniquelyIdentified: Bool {
+    switch baseKind {
+    case .box:
+      // The operand `%op` in `%baseAddress = project_box %op` can
+      // be `alloc_box` or an argument. Only if it's a fresh allocation it is
+      // uniquelyIdentified, otherwise it is aliasable, as all the other references.
+      return baseAddress is AllocBoxInst
+    case .stack, .global:
+      return true
+    case .argument:
+      // An argument address that is non-aliasable
+      return (baseAddress as! FunctionArgument).isExclusiveIndirectParameter
+    case .class, .tail, .yield, .pointer:
+      // References (.class and .tail) may alias, and so do pointers and
+      // yield results
+      return false
+    }
+  }
+
+  /// Returns `true` if the two access bases do not alias
```

Comment:
nitpick: maybe it's better to invert this function: `func mayAlias(with:) -> Bool`